File: OpenX/python/openx/core/engine.py
from __future__ import annotations
import os, sys
from enum import Enum, Flag
from openx.core.core import ISource
from openx.core.util import decodePath, YamlDict

#region Engine

# Engine
class Engine:
    def __init__(self, id: str, name: str):
        self.enabled: bool = True
        self.caps: EngineX.Caps = EngineX.Caps.None_
        self.id: str = id
        self.name: str = name
        self.tag: str = None
        self.gfxFactory: callable = None
        self.sfxFactory: callable = None
        self.logFunc: callable = lambda a: print(a)
    def activate(self) -> None: pass
    def deactivate(self) -> None: pass

# EngineX
class EngineX:
    # The Caps.
    class Caps(Flag):
        None_ = 0x0
        ReadDds = 0x1

    # The OS.
    class OS(Enum):
        Unknown = 0
        Windows = 1
        OSX = 2
        Linux = 3
        Android = 4

    @staticmethod
    def activate(engine: Engine) -> None:
        if not engine or not engine.enabled: engine = UnknownEngine.this
        EngineX.engines.add(engine)
        current = EngineX.current
        if current != engine:
            if current: current.deactivate()
            if engine: engine.activate()
            EngineX.gfx = engine.gfxFactory() if engine and engine.gfxFactory else None
            EngineX.sfx = engine.sfxFactory() if engine and engine.sfxFactory else None
            EngineX.current = engine
        return engine

    @staticmethod
    def createMatcher(searchPattern: str) -> callable:
        if not searchPattern: return lambda x: True
        wildcardCount = searchPattern.count('*')
        if wildcardCount <= 0: return lambda x: x.casefold() == searchPattern.casefold()
        elif wildcardCount == 1:
            newPattern = searchPattern.replace('*', '')
            if searchPattern.startswith('*'): return lambda x: x.casefold().endswith(newPattern)
            elif searchPattern.endswith('*'): return lambda x: x.casefold().startswith(newPattern)
        regexPattern = f'^{re.escape(searchPattern).replace('\\*', '.*')}$'
        @staticmethod
        def _lambdax(x: str):
            try: return re.match(x, regexPattern)
            except: return False
        return _lambdax

    @staticmethod
    def decodePath(path: str, rootPath: str = None) -> str: return decodePath(EngineX.applicationPath, path, rootPath)

    platformOS: OS = OS.Windows if sys.platform == 'win32' else \
        OS.OSX if sys.platform == 'darwin' else \
        OS.Linux if sys.platform.startswith('linux') else \
        OS.Unknown
    engines: set[object] = {}
    inTestHost: bool = 'unittest' in sys.modules.keys()
    applicationPath = os.getcwd()
    options = YamlDict('~/.gamex.yaml')
    current: Engine = None
    gfx: list[IOpenGfx] = None
    sfx: list[IOpenSfx] = None

#endregion

from openx.engines.test import TestEngine
from openx.engines.unknown import UnknownEngine

EngineX.engines = { UnknownEngine.this }
EngineX.current = EngineX.activate(TestEngine.this if EngineX.inTestHost else UnknownEngine.this)

File: OpenX/dotnet/Core/OpenX/Engine.cs
using OpenX.Gfx;
using OpenX.Sfx;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;

namespace OpenX;

#region Engine

/// <summary>
/// Gets the engine.
/// </summary>
public abstract class Engine(string id, string name) {
    /// <summary>
    /// Gets the engine enabled.
    /// </summary>
    public bool Enabled = true;

    /// <summary>
    /// Gets the engine caps
    /// </summary>
    public EngineX.Caps Caps = EngineX.Caps.None_;

    /// <summary>
    /// Gets the engine id.
    /// </summary>
    public readonly string Id = id;

    /// <summary>
    /// Gets the engine name.
    /// </summary>
    public readonly string Name = name;

    /// <summary>
    /// Gets the engine name.
    /// </summary>
    public string DisplayName => Name;

    /// <summary>
    /// Gets the engine tag.
    /// </summary>
    public string Tag;

    /// <summary>
    /// Gets the engine gfx factory.
    /// </summary>
    public Func<IOpenGfx[]> GfxFactory = () => null; // throw new Exception("No GfxFactory");

    /// <summary>
    /// Gets the engine sfx factory.
    /// </summary>
    public Func<IOpenSfx[]> SfxFactory = () => null; // throw new Exception("No SfxFactory");

    /// <summary>
    /// Gets the engine log func.
    /// </summary>
    public Action<string> LogFunc = a => System.Diagnostics.Debug.Print(a);

    /// <summary>
    /// Activates the engine.
    /// </summary>
    public virtual void Activate() {
        Log.Func = LogFunc;
    }

    /// <summary>
    /// Deactivates the engine.
    /// </summary>
    public virtual void Deactivate() { }
}

/// <summary>
/// EngineX
/// </summary>
public static class EngineX {
    public static readonly float Epsilon = GetEpsilon();
    static float GetEpsilon() { float epsilon = 1f, comparison; do { epsilon *= 0.5f; comparison = 1.0f + epsilon; } while (comparison > 1.0f); return epsilon; }
    //public static Action Hook;

    /// <summary>
    /// The platform Caps.
    /// </summary>
    [Flags] public enum Caps { None_ = 0x0, ReadDds = 0x1, Drawing = 0x2 }

    /// <summary>
    /// The platform OS.
    /// </summary>
    public enum OS { Unknown, Windows, OSX, Linux, Android }

    /// <summary>
    /// Gets the platform os.
    /// </summary>
    public static readonly OS PlatformOS = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? OS.Windows
        : RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? OS.OSX
        : RuntimeInformation.OSDescription.StartsWith("android-") ? OS.Android
        : RuntimeInformation.IsOSPlatform(OSPlatform.Linux) ? OS.Linux
        : OS.Unknown;
    //: throw new ArgumentOutOfRangeException(nameof(RuntimeInformation.IsOSPlatform), RuntimeInformation.OSDescription);

    /// <summary>
    /// Gets the platform startups.
    /// </summary>
    public static readonly HashSet<Engine> Engines = [UnknownEngine.This];

    /// <summary>
    /// Determines if in a test host.
    /// </summary>
    public static readonly bool InTestHost = AppDomain.CurrentDomain.GetAssemblies().Any(x => x.FullName.StartsWith("testhost,"));

    /// <summary>
    /// Gets the application Path
    /// </summary>
    public static readonly string ApplicationPath = AppDomain.CurrentDomain.BaseDirectory;

    /// <summary>
    /// Gets the platform startups.
    /// </summary>
    public static YamlDict Options = new("~/.gamex.yaml");

    /// <summary>
    /// Gets or sets the current platform.
    /// </summary>
    public static Engine Current = Activate(InTestHost ? TestEngine.This : UnknownEngine.This);

    /// <summary>
    /// Activates an engine.
    /// </summary>
    public static Engine Activate(Engine engine) {
        //Hook?.Invoke(); Hook = null;
        if (engine == null || !engine.Enabled) engine = UnknownEngine.This;
        Engines.Add(engine);
        var current = Current;
        if (current != engine) {
            current?.Deactivate();
            engine?.Activate();
            Gfx = engine?.GfxFactory?.Invoke();
            Sfx = engine?.SfxFactory?.Invoke();
            Current = engine;
        }
        return engine;
    }

    /// <summary>
    /// Gets the gfx.
    /// </summary>
    /// <value>
    /// The gfx.
    /// </value>
    public static IOpenGfx[] Gfx { get; internal set; } = null;

    /// <summary>
    /// Gets the gfx.
    /// </summary>
    /// <value>
    /// The sfx.
    /// </value>
    public static IOpenSfx[] Sfx { get; internal set; } = null;

    ///// <summary>
    ///// Creates the matcher.
    ///// </summary>
    ///// <param name="searchPattern">The searchPattern.</param>
    ///// <returns></returns>
    //public static Func<string, bool> CreateMatcher(string searchPattern) {
    //    if (string.IsNullOrEmpty(searchPattern)) return x => true;
    //    var wildcardCount = searchPattern.Count(x => x.Equals('*'));
    //    if (wildcardCount <= 0) return x => x.Equals(searchPattern, StringComparison.CurrentCultureIgnoreCase);
    //    else if (wildcardCount == 1) {
    //        var newPattern = searchPattern.Replace("*", "");
    //        if (searchPattern.StartsWith("*")) return x => x.EndsWith(newPattern, StringComparison.CurrentCultureIgnoreCase);
    //        else if (searchPattern.EndsWith("*")) return x => x.StartsWith(newPattern, StringComparison.CurrentCultureIgnoreCase);
    //    }
    //    var regexPattern = $"^{Regex.Escape(searchPattern).Replace("\\*", ".*")}$";
    //    return x => {
    //        try { return Regex.IsMatch(x, regexPattern); }
    //        catch { return false; }
    //    };
    //}

    public static string DecodePath(string path, string rootPath = null) => Util.DecodePath(ApplicationPath, path, rootPath);
}

#endregion
